db/write_thread.cc

```diff
@@ -303,12 +330,50 @@ void WriteThread::CompleteFollower(Writer* w, WriteGroup& write_group) {
   SetState(w, STATE_COMPLETED);
 }
 
+void WriteThread::BeginWriteStall() {
+  LinkOne(&write_stall_dummy_, &newest_writer_, false);
+
+  // Walk writer list until w->write_group != nullptr. The current write group
+  // will not have a mix of slowdown/no_slowdown, so its ok to stop at that
+  // point
+  Writer* w = write_stall_dummy_.link_older;
+  Writer* prev = &write_stall_dummy_;
+  while (w != nullptr && w->write_group == nullptr) {
+    if (w->no_slowdown) {
+      prev->link_older = w->link_older;
+      w->status = Status::Incomplete();
+      SetState(w, STATE_COMPLETED);
+      w = prev->link_older;
+    } else {
+      prev = w;
+      w = w->link_older;
+    }
+  }
+}
+
+void WriteThread::EndWriteStall() {
+  std::lock_guard<std::mutex> lock(mu_);
+
+  assert(newest_writer_.load(std::memory_order_relaxed) == &write_stall_dummy_);
+  Writer* expected = &write_stall_dummy_;
+  newest_writer_.compare_exchange_weak(expected, write_stall_dummy_.link_older);
```

Comment:
`compare_exchange_weak` can fail for no reason, use `compare_exchange_strong` at least. And I think it is safe to use `exchange` here, since no other thread will perform a write to the linked list if they see the dummy.